### PR TITLE
Add KMS<->DB abstraction

### PR DESCRIPTION
**Pull request by sethvargo**
- Fixes GH-165
- Fixes GH-117

**Release Note**

```release-note
Use Key Manager to encrypt SMS configurations
```

### CODE REVIEW

**Comment by sethvargo** [comment]
screenshot-20200805-212541@2x

File: cmd/server/assets/realm.html [hunk 1]
```diff
@@ -18,15 +18,57 @@ <h1>Realm settings</h1>
```
Comment (sethvargo) [inline cmd/server/assets/realm.html:18]: See screenshot for render

File: pkg/database/sms_config.go [hunk 2]
```diff
@@ -16,58 +16,113 @@ package database
...
+	if r.twilioAuthToken != "" {
+		r.TwilioAuthToken = r.twilioAuthToken
+	}
+	return nil
+}
+
+// AfterFind runs after a record is found. It's used to mutate values (such as
+// auth tokens) before storing them on the struct. Do not call this function,
+// gorm calls it automatically.
+func (r *SMSConfig) AfterFind() error {
```
Comment (mikehelmick) [inline pkg/database/sms_config.go:92]: There needs to be some kind of auto upgrade from secret to encrypted in DB for existing installations.

Comment (sethvargo) [reply pkg/database/sms_config.go:93]: Added

File: pkg/controller/realmadmin/save.go [hunk 3]
```diff
@@ -48,14 +56,74 @@ func (c *Controller) HandleSave() http.Handler {
 			return
 		}
 
+		// SMS config is all or nothing
+		if (form.TwilioAccountSid != "" || form.TwilioAuthToken != "" || form.TwilioFromNumber != "") &&
+			(form.TwilioAccountSid == "" || form.TwilioAuthToken == "" || form.TwilioFromNumber == "") {
+			flash.Error("Error updating realm: all SMS fields are required")
```
Comment (mikehelmick) [inline pkg/controller/realmadmin/save.go:62]: nit: not quite accurate, either they're all required or must be blank to disable SMS

Comment (sethvargo) [reply pkg/controller/realmadmin/save.go]: Done

**Pull request merged**